### ISSUE TITLE
sshd: IndexError: list index out of range

### ISSUE THREAD

**Issue by sean797** [FrostyX/tracer#109]
Created from https://github.com/FrostyX/tracer/issues/100#issuecomment-362924861

----
@Wrecknor

/usr/lib64/python3.6/re.py:212: FutureWarning: split() requires a non-empty pattern match.
return _compile(pattern, flags).split(string, maxsplit)

Tracer:
Call to Tracer API ended unexpectedly:

Traceback (most recent call last):
File "/usr/lib/python3.6/site-packages/dnf-plugins/tracer.py", line 67, in transaction
tracer.render()
File "/usr/lib/python3.6/site-packages/dnf-plugins/tracer.py", line 92, in render
return view.render()
File "/usr/lib/python3.6/site-packages/tracer/views/default.py", line 59, in render
{"title": " * " + _("Some applications using:"), "content": with_helpers_content()},
File "/usr/lib/python3.6/site-packages/tracer/views/default.py", line 20, in with_helpers_content
applications = self.args.applications.with_helpers().exclude_types(types).sorted("helper")
File "/usr/lib/python3.6/site-packages/tracer/resources/collections.py", line 38, in with_helpers
return ApplicationsCollection(applications)
File "/usr/lib/python3.6/site-packages/tracer/resources/collections.py", line 37, in
applications = filter(lambda app: app.helper, self)
File "/usr/lib/python3.6/site-packages/tracer/resources/applications.py", line 217, in helper
helper = self._attributes["helper"] if self._attributes["helper"] else Applications._helper(self)
File "/usr/lib/python3.6/site-packages/tracer/resources/applications.py", line 120, in _helper
if app.type == Applications.TYPES["DAEMON"]:
File "/usr/lib/python3.6/site-packages/tracer/resources/applications.py", line 202, in type
elif self.is_session:
File "/usr/lib/python3.6/site-packages/tracer/resources/applications.py", line 194, in is_session
if re.search('ssh-.*-session', str(self.name)):
File "/usr/lib/python3.6/site-packages/tracer/resources/applications.py", line 270, in name
if self.instances and bus.unit_path_from_pid(self.instances[0].pid):
File "/usr/lib/python3.6/site-packages/tracer/resources/applications.py", line 260, in instances
lambda process: self._attributes["name"] in [process.name(), process.real_name])
File "/usr/lib/python3.6/site-packages/tracer/resources/collections.py", line 83, in filtered
if function(process):
File "/usr/lib/python3.6/site-packages/tracer/resources/applications.py", line 260, in
lambda process: self._attributes["name"] in [process.name(), process.real_name])
File "/usr/lib/python3.6/site-packages/tracer/resources/processes.py", line 73, in name
return 'ssh-{0}-session'.format(re.split(' |@|',self._attr("cmdline")[0])[1])
IndexError: list index out of range

Please visit https://github.com/FrostyX/tracer/issues and submit the issue. Thank you
We apologize for any inconvenience

------

@elyscape 

So I did some digging and here's what's going on, at least some of the time:
```shellsession
$ pgrep -a sshd
744 /usr/sbin/sshd -D
9641 sshd: elyscape [priv]
9651 sshd: elyscape@pts/0
```
If a `ProcessWrapper` is created for process 9641, here's the logic that happens when its `name` method is called, based on [this code][pw.name]:

71. `self._attr("name")` returns `'sshd'`
72. `self._attr("exe")` returns `'/usr/bin/sshd'`, which is not in `'sshd'`
73. Oneliner hell:
    1. `self._attr("cmdline")` returns `['sshd:', 'elyscape', '[priv]']`
    2. `self._attr("cmdline")[0]` is `'sshd:`
    3. `re.split()` is called with the regular expression ` |@|` (which results in a deprecation error due to the empty match and should probably just be ` |@` or `[ @]`) and the search string `'sshd:'`
    4. This returns `['sshd:']`
    5. `re.split()[1]` throws an IndexError

The easiest way to fix this is probably to just change line 73 to run the regular expression on `" ".join(self._attr("cmdline"))` (i.e. join the contents of `self._attr("cmdline")` with spaces into a single string) instead of using the first element. In that case, the flow would look like this:

71. `self._attr("name")` returns `'sshd'`
72. `self._attr("exe")` returns `'/usr/bin/sshd'`, which is not in `'sshd'`
73. Oneliner hell:
    1. `self._attr("cmdline")` returns `['sshd:', 'elyscape', '[priv]']`
    2. `" ".join(self._attr("cmdline"))` is `'sshd: elyscape [priv]'`
    3. `re.split()` is called with the regular expression ` |@` and the search string `'sshd: elyscape [priv]'`
    4. This returns `['sshd:', 'elyscape', '[priv]']`
    5. `re.split()[1]` is `'elyscape'`
    6. `'ssh-{0}-session'.format()` is called with `'elyscape'`
    7. this returns `'ssh-elyscape-session'`
    8. `ProcessWrapper.name()` returns that

[pw.name]: https://github.com/FrostyX/tracer/blob/tracer-0.6.13.1-1/tracer/resources/processes.py#L67-L76

**Comment by elyscape** [comment]
This particular case should have been fixed by #106.

**Comment by sean797** [comment]
Okay - I was look into it, but couldn't reproduce - that makes sense. Thanks!
